libs/custom-paper-handler/src/printing.ts

```diff
@@ -0,0 +1,220 @@
+  whiteThreshold: DEFAULT_GRAYSCALE_WHITE_THRESHOLD,
+};
+
+/**
+ * Converts 8-bit sRGB color values to a binary black/white representation.
+ * Uses weighted method without gamma correction for speed.
+ *
+ * @param r Red color value from 0 - 255
+ * @param g Green color value from 0 - 255
+ * @param b Blue color value from 0 - 255
+ * @returns true for black, false for white
+ */
+function rgbToBinary(
+  r: number,
+  g: number,
+  b: number,
+  options: ImageConversionOptions
+): boolean {
+  const grayscaleValue = (
+    options.useGammaConversion ? rgbToGrayscaleGamma : rgbToGrayscale
+  )(r, g, b);
+  return grayscaleValue < options.whiteThreshold;
+}
+
+export function imageDataToBinaryBitmap(
+  imageData: ImageData,
+  overrideOptions: Partial<ImageConversionOptions> = {}
+): BinaryBitmap {
+  const options: ImageConversionOptions = {
+    ...DEFAULT_IMAGE_CONVERSION_OPTIONS,
+    ...overrideOptions,
+  };
+
+  const data: boolean[] = [];
+
+  let r = 0;
+  let g = 0;
+  let b = 0;
+  imageData.data.forEach((element, index) => {
+    // ImageData.data is in RGBA format. Map RBG values to grayscale.
+    // eslint-disable-next-line default-case
+    switch (index % 4) {
+      case 0:
+        r = element;
+        return;
+      case 1:
+        g = element;
+        return;
+      case 2:
+        b = element;
+        return;
+      case 3:
+        data.push(rgbToBinary(r, g, b, options));
+    }
+  });
+
+  return {
+    data,
+    width: imageData.width,
+    height: imageData.height,
+  };
+}
+
+export function chunkBinaryBitmap(
+  binaryBitmap: BinaryBitmap
+): PaperHandlerBitmapExt[] {
+  if (binaryBitmap.width > 1600) {
+    throw new Error('binaryBitmap width exceeds printers allowed size');
+  }
+
+  const paperHandlerBitmaps: PaperHandlerBitmapExt[] = [];
+
+  // Each chunk will be 24 dots high. Since for this prototype, we're likely
+  // not printing in the lowest 8 or 24 rows of dots, just ignore those.
+  const numChunkRows = Math.floor(binaryBitmap.height / 24);
+  for (
+    let chunkRowIndex = 0;
+    chunkRowIndex < numChunkRows;
+    chunkRowIndex += 1
+  ) {
+    const chunkOrderBits: boolean[] = [];
+    let empty = true;
+    for (let column = 0; column < binaryBitmap.width; column += 1) {
+      for (
+        let row = chunkRowIndex * 24;
+        row < chunkRowIndex * 24 + 24;
+        row += 1
+      ) {
+        const bit = binaryBitmap.data[row * binaryBitmap.width + column];
+        assert(bit !== undefined);
+        chunkOrderBits.push(bit);
+        if (bit) {
+          empty = false;
+        }
+      }
+    }
+
+    if (empty) {
+      paperHandlerBitmaps.push({
+        data: new Uint8Array([]),
+        width: binaryBitmap.width,
+        empty,
+      });
+      continue;
+    }
+
+    const chunks: number[] = [];
+
+    for (let i = 0; i < chunkOrderBits.length / 8; i += 1) {
+      chunks.push(
+        bitArrayToByte(chunkOrderBits.slice(i * 8, i * 8 + 8) as BitArray)
+      );
+    }
```

Comment:
This feels like it could be pretty slow. Something to keep an eye on and measure.